lib/hw_id/hw_id.c

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <hw_id.h>
+#include <zephyr/kernel.h>
+
+/* includes for the different HW ID sources */
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_BLE_MAC)
+#include <zephyr/bluetooth/bluetooth.h>
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_BLE_MAC) */
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_DEVICE_ID)
+#include <zephyr/drivers/hwinfo.h>
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_DEVICE_ID) */
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_UUID)
+#include <modem/modem_jwt.h>
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_UUID) */
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_IMEI)
+#include <nrf_modem_at.h>
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_IMEI) */
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_NET_MAC)
+#include <zephyr/net/net_if.h>
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_NET_MAC) */
+
+#define IMEI_LEN 15
+
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_BLE_MAC)
+/* Ask BLE stack for default MAC address */
+
+int hw_id_get(char *buf, size_t buf_len)
+{
+	if (buf == NULL || buf_len < HW_ID_LEN) {
+		return -EINVAL;
+	}
+
+	struct bt_le_oob oob;
+	int ret = bt_le_oob_get_local(BT_ID_DEFAULT, &oob);
+
+	if (ret) {
+		return -EIO;
+	}
+
+	snprintk(buf, buf_len, "%02X%02X%02X%02X%02X%02X",
+		oob.addr.a.val[5],
+		oob.addr.a.val[4],
+		oob.addr.a.val[3],
+		oob.addr.a.val[2],
+		oob.addr.a.val[1],
+		oob.addr.a.val[0]);
+	return 0;
+}
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_BLE_MAC) */
+
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_DEVICE_ID)
+/* Directly read Device ID from registers */
+
+int hw_id_get(char *buf, size_t buf_len)
+{
+	if (buf == NULL || buf_len < HW_ID_LEN) {
+		return -EINVAL;
+	}
+
+	uint8_t device_id[8];
+	ssize_t ret = hwinfo_get_device_id(device_id, ARRAY_SIZE(device_id));
+
+	if (ret != ARRAY_SIZE(device_id)) {
+		return -EIO;
+	}
+
+	snprintk(buf, buf_len, "%02X%02X%02X%02X%02X%02X%02X%02X",
+		device_id[0],
+		device_id[1],
+		device_id[2],
+		device_id[3],
+		device_id[4],
+		device_id[5],
+		device_id[6],
+		device_id[7]);
+	return 0;
+}
+#endif /* defined(CONFIG_HW_ID_LIBRARY_SOURCE_DEVICE_ID) */
+
+#if defined(CONFIG_HW_ID_LIBRARY_SOURCE_UUID)
+/* Request a UUID from the modem */
+
+int hw_id_get(char *buf, size_t buf_len)
+{
+	if (buf == NULL || buf_len < HW_ID_LEN) {
+		return -EINVAL;
+	}
+
+	struct nrf_device_uuid dev = {0};
+
+	int err = modem_jwt_get_uuids(&dev, NULL);
```

Comment:
What's the difference between a UUID from `jwt` and the modem firmware UUID from the AT command?

Reply:
Are you referring to `%XMODEMUUID`?  
That AT command returns the UUID of the _installed modem firmware_.
The UUID of interest here is the _device UUID_, which can be found in the "iss" claim of a JWT.

AFAIK, the only other way to obtain the device UUID is by parsing an attestation token `%ATTESTTOKEN`.